docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+```sql
+-- Generate random op_id.
+BEGIN;
+-- Get the latest state (with the largest sequence number).
+SELECT FOR UPDATE * FROM tenant_usage WHERE tenant_id=.. ORDER BY seq DESC LIMIT 1;
+-- Calculate new state.
+-- Set the new state, using the next sequence number.
+INSERT INTO tenant_usage (tenant_id, seq, ...) (tenant_id, seq+1, ...);
+END;
+```
+
+Any failed request is retried with the same `op_id`. If the previous request
+went through, the transaction will error out with a conflict on `op_id`,
+allowing special handling of this case.
+
+We perform periodic garbage collection to keep the table from growing. Each
+operation can trigger garbage collection (with random chance).
+
+```sql
+-- by number of entries
+DELETE FROM tenant_usage WHERE tenant_id=.. AND seq < (
+  SELECT seq - 1000 FROM tenant_usage WHERE tenant_id=.. ORDER BY seq DESC LIMIT 1
+)
+-- or by timestamp:
+DELETE FROM tenant_usage WHERE tenant_id=.. AND op_time < now() - '24hr'
+```
+
+TODO: it's unfortunate that we will have to talk to two ranges on each
+operation. Could we achieve this with a single index? We could reserve part of
+the UUID space for the sequence number, allowing one index to serve a
+dual-purpose, but things would become messy.
+
+Another possible variation would be to have a single `(tenant_id, op_id)`
+primary index and *also* store the current state and sequence number in
+`(tenant_id, '00000000-0000-0000-0000-000000000000')`. That would not require a
+secondary index, but would generate larger KVs as we store all fields twice.
+
+## KV Connector API
+
+The KV Connector API is used by the SQL pods to make requests to the distributed
+bucket. This API is implemented on the host cluster side by performing a
+transaction on `tenant_usage`.
+
+```go
+type TenantTokenBucketProvider interface {
+   Request(
+     ctx context.Context,
+     uuid util.UUID,
+     prevShareValue float64,
+     prevShareTimestamp time.Time,
+     tokensRequested float64,
+     ruUsage float64,
+     readRequests int64,
+     readBytes int64,
+     writeRequests int64,
+     writeBytes int64,
+     totalPodCPUSeconds float64,
+   ) (
+     updateTimestamp time.Time,
+     grantedTokens float64,
+     grantedTimestamp time.Time,
+     maxBurstTokens float64
+   )
+```
+
+
+## Configuration API
+
+The limits for a tenant are periodically reconfigured. The proposed interface is
+via an internal SQL function; we can easily provide other endpoints as
+necessary.
+
+```sql
+SELECT crdb_internal.update_tenant_resource_limits(
+  tenant_id,                    -- INT
+  operation_uuid,               -- UUID
+  as_of,                        -- TIMESTAMP
+  as_of_consumed_request_units, -- FLOAT
+  available_request_units,      -- FLOAT
+  refill_rate,                  -- FLOAT
+  max_request_units             -- FLOAT
+)
+```
+
+The statement completes after a new row is inserted in the system table.
+
+Updated settings are calculated based on the total consumption reported at some
```

Comment:
This is a bit terse, I feel like this is leaving the reader to figure out what's actually getting reconfigured here and which parts match what happens on a regular acquisition (i.e. why do I need to pass a uuid, what's as_of, etc)